---
title: "Plot proportion of indirect effects vs cis hit ratio"
author: "Andreas Gschwind"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
---

## Goal
Calculate indirect effects positive ratio and compare to cis-effect positive ratio as a function of
distance to TSS.

```{r requiredPackages, warning=FALSE, message=FALSE}
# save.image(paste0("analyze_", snakemake@wildcards$sample, ".rda"))
# stop()
library(tidyverse)
library(data.table)
library(Matrix)
```

## Used data
Results from differential expression tests for trans-acting effects are used to estimate the
proportion of indirect effects in CRISPR data. Proportion of indirect effects is compared to
proportion of direct cis effects from main differential expression results.
```{r}
# load trans effect results
trans_results <- readRDS(snakemake@input$trans_results)

# load cis effects results from previous run
cis_results <- read_tsv(snakemake@input$cis_results, show_col_types = FALSE)

# load processed ENCODE format results including filters
encode_results <- read_tsv(snakemake@input$encode_file, show_col_types = FALSE)

# load gene universe file
gene_univ <- read_tsv(snakemake@input$gene_univ, show_col_types = FALSE)

# load gene expression data
gene_expr <- read_csv(snakemake@input$gene_expr, show_col_types = FALSE)
```

## Define significan trans-acting effects
Define significant trans effects by applying the same p-value threshold as for the cis effects.
```{r}
# significance threshold for cis effects
pval_threshold_cis <- cis_results %>% 
  filter(significant == TRUE) %>% 
  pull(p_value) %>% 
  max()

# add significance to trans-effect results based on cis significance threshold
trans_results <- trans_results %>% 
  mutate(significant = p_value <= pval_threshold_cis,
         regulated = significant & log_2_fold_change < 0)
```

## Filter data
Filter trans and cis effects based on ValidConnection in ENCODE format data. For trans effects, any
perturbations that are TSS controls or overlap potential promoters are removed. For cis effects,
the same filter is applied in addition to removing any E-G pairs where the perturbation is within
the gene body of their target genes. These filters consistent with all other ENCODE distal
regulation analyses.
```{r}
# get gene symbol for every gene id from processed ENCODE data
genes <- encode_results %>%
  select(gene = measuredEnsemblID, gene_symbol = measuredGeneSymbol) %>% 
  distinct()

# add gene symbols to cis results 
cis_results <- cis_results %>% 
  left_join(genes, by = c("response_id" = "gene")) %>% 
  filter(!is.na(gene_symbol))

# add ValidConnection column to cis results
cis_results <- cis_results %>% 
  mutate(name = paste0(gene_symbol, "|", grna_target, ":.")) %>% 
  left_join(distinct(select(encode_results, name, ValidConnection)), by = "name")
```

```{r}
# get valid enhancers only, which also removes any TSS controls
enh_filter <- c("overlaps potential promoter", "TSS targeting guide(s)")
valid_enh <- encode_results %>% 
  filter(!ValidConnection %in% enh_filter) %>% 
  pull(PerturbationTargetID)

# add ValidConnection column to trans-results
trans_results <- trans_results %>% 
  mutate(pert_id = paste0(grna_target, ":.")) %>% 
  mutate(ValidConnection = if_else(pert_id %in% valid_enh, true = "TRUE", false = "Invalid enhancer"))
```

```{r}
# filter cis and trans results based on ValidConnection
cis_results_filt <- filter(cis_results, ValidConnection == "TRUE")
trans_results_filt <- filter(trans_results, ValidConnection == "TRUE")

# only retain data on genes in distal regulation genes universe
cis_results_filt <- filter(cis_results_filt, response_id %in% gene_univ$Ensembl_ID)
trans_results_filt <- filter(trans_results_filt, response_id %in% gene_univ$Ensembl_ID)
```

## Calculate positive ratios
Calculate proportion of positives for both trans effects and cis effects. For the latter, the
proportion of positives is calculates as a function of distance to TSS.

Bin cis results into 50 bins based on distance to TSS and compute proportion of positives for each
bin:
```{r}
# add regulated columns to cis results
cis_results_filt <- cis_results_filt %>% 
  mutate(regulated = significant & log_2_fold_change < 0) %>% 
  filter(!is.na(regulated))

# bin cis pairs by distance
cis_results_filt <- cis_results_filt %>% 
  left_join(select(encode_results, name, distToTSS), by = "name") %>% 
  filter(distToTSS <= 2e6) %>% 
  mutate(dist_bin = cut(distToTSS, breaks = 25))

# calculate positive ratios and mean distance per distance bin
cis_pos_ratio <- cis_results_filt %>% 
  group_by(dist_bin) %>% 
  summarize(positive_ratio = mean(regulated),
            mean_dist = mean(distToTSS)) %>% 
  mutate(Type = "Cis-interactions")
```

Calculate proportion of positives for trans effects regardless of distance (all are on other
chromosomes than the perturbation).
```{r}
# calculate trans-effects positive ratio
trans_pos_ratio <- mean(trans_results_filt$regulated, na.rm = TRUE)

# create mock table
trans_pos_ratio_table <- tibble(mean_dist = c(0, 2e6),
       positive_ratio = trans_pos_ratio,
       Type = "Trans-interactions")
```

Based on this analysis, the expected trans-effect ratio is: `r trans_pos_ratio`.

## Plot cis vs trans positive ratios
Plot the proportion of positives for cis effects as a function of distance to TSS. Add a horizontal
line to indicate the expected proportion of indirect effects estimated from testing for trans
effects. The red line indicates where the proportion of cis effects starts to be equal to the
expected proportion of indirect effects. 
```{r, fig.height=3.5, fig.width=7}
# manually inferred distance to TSS threshold where proportion of indirect effects is same as
# proportion of effects in cis
dist_threshold <- snakemake@params$dist_filter

# plot proportion of positives as function of distance
plot_title <- paste(snakemake@wildcards$sample, "cis vs. trans positive ratio")
ggplot(cis_pos_ratio, aes(x = mean_dist, y = positive_ratio, color = Type)) +
  geom_point() +
  geom_line(data = trans_pos_ratio_table) +
  geom_vline(xintercept = dist_threshold, color = "firebrick") +
  labs(x = "Distance to TSS (0-2Mb, 25 bins)", y = "Proportion positives", title = plot_title) +
  scale_color_manual(values = c("black", "dodgerblue3")) +
  theme_bw()
```

## Calculate positive ratios beyond distance threshold
```{r}
# function to calculate hit ratios within a given distance to tss range
calc_hit_ratios_dist <- function(cis_results, trans_pos_ratio, min_dist, max_dist) {
  
  # get total number of cis-hits below max distance
  cis_results <- filter(cis_results, distToTSS <= max_dist)
  total_cis_hits <- sum(cis_results$regulated)
  
  # calculate proportions of hits above min distance
  cis_results %>% 
    filter(distToTSS > min_dist) %>% 
    summarize(far_cis_hits = sum(regulated),
              far_cis_pos_ratio = mean(regulated)) %>% 
    mutate(trans_pos_ratio = trans_pos_ratio,
           far_cis_vs_trans_ratio = far_cis_pos_ratio / trans_pos_ratio,
           expect_true_far_cis_prop = far_cis_pos_ratio / (far_cis_pos_ratio + trans_pos_ratio)) %>%
    mutate(total_cis_hits = total_cis_hits, .before = far_cis_hits) %>% 
    mutate(prop_far_cis_hits = far_cis_hits / total_cis_hits, .after = far_cis_hits)
}

# compute overall cis-hit ratio beyond 250kb
cis_stats_250kb_2mb <- calc_hit_ratios_dist(cis_results_filt, trans_pos_ratio = trans_pos_ratio,
                                            min_dist = dist_threshold, max_dist = 2e6)

# compute overall cis-hit ratio beyond 250kb and within 1Mb
cis_stats_250kb_1mb <- calc_hit_ratios_dist(cis_results_filt, trans_pos_ratio = trans_pos_ratio,
                                            min_dist = dist_threshold, max_dist = 1e6)

# combine into one table
cis_stats <- bind_rows(`2mb` = cis_stats_250kb_2mb, `1mb` = cis_stats_250kb_1mb, .id = "max_dist")

# save to table
outfile <- file.path("results", snakemake@wildcards$sample, "indirect_effects_proportions.csv")
write_csv(cis_stats, file = outfile)
```

## Compare gene expression distributions
Compare expression distributions between genes in cis- and trans-acting pairs.
```{r, fig.width=4, fig.height=4.5}
# add gene expression levels to sceptre results
trans_results_filt <- left_join(trans_results_filt, gene_expr, by = c("response_id" = "gene"))
cis_results_filt <- left_join(cis_results_filt, gene_expr, by = c("response_id" = "gene"))

# get expression for genes in cis- and trans-pairs
trans_filt_expr <- select(trans_results_filt, response_id, avg_expr)
cis_filt_expr <- select(cis_results_filt, response_id, avg_expr)
expr <- bind_rows(trans = trans_filt_expr, cis = cis_filt_expr, .id = "type")

# only retain unique genes in both trans and cis gene lists
expr <- distinct(expr)

# plot expression distributions between genes in cis- and trans-pairs
ggplot(expr, aes(x = type, y = avg_expr, fill = type)) +
  geom_violin() +
  geom_boxplot(width = 0.25) +
  labs(title = paste("Gene expression", snakemake@wildcards$sample ), y = "Avg. UMI per cell",
       x = "Analysis") +
  scale_fill_manual(values = c(cis = "steelblue", trans = "goldenrod")) +
  scale_y_log10() +
  theme_bw() +
  theme(legend.position = "none")
```

## Create filtered benchmarking dataset
If a benchmarking dataset is available for this dataset, filter it based on distance to TSS to remove
positives that are more likely to be indirect effects. Also remove an equal proportion of randomly
selected negatives to maintain class balance for benchmarking.
```{r}
if (!is.null(snakemake@input$ep_benchmarking_file)) {
  
  # load dataset in EPbenchmarking format
  crispr_file <- snakemake@input$ep_benchmarking_file
  crispr <- read_tsv(crispr_file, show_col_types = FALSE)
  
  # extract original column names for later formatting
  crispr_cols <- colnames(crispr)
  
  # calculate distance to TSS
  crispr <- crispr %>%
    mutate(distToTSS = if_else(
      chrom == chrTSS,
      true = abs(((chromStart + chromEnd) / 2) - ((startTSS + endTSS) / 2)),
      false = NA_real_))
  
  # get all positive and negatives
  positives <- filter(crispr, Regulated == TRUE)
  negatives <- filter(crispr, Regulated == FALSE)
  
  # filter positives based on distance threshold and get proportion of filtered pairs
  positives_filt <- filter(positives, distToTSS <= dist_threshold)
  filt_prop <- nrow(positives_filt) / nrow(positives)
  
  # randomly sample an equal proportion of negatives
  negatives_sample <- sample_n(negatives, size = round(nrow(negatives) * filt_prop))
  
  # combine filtered positives and sampled negatives to form output
  crispr_prop_filt <- bind_rows(positives_filt, negatives_sample) %>% 
    arrange(chrom, chromStart, chromEnd, measuredGeneSymbol) %>% 
    select(all_of(crispr_cols))
  
  # save to output file
  outfile <- paste0(tools::file_path_sans_ext(basename(crispr_file)), "_propFilter.tsv.gz")
  outfile <- file.path(dirname(snakemake@input$trans_results), outfile)
  write_tsv(crispr_prop_filt, file = outfile)

}
```
